backend/de.metas.business/src/main/java/de/metas/fulltextsearch/indexer/handler/invoice/InvoiceFTSModelIndexer.java

```diff
@@ -0,0 +1,260 @@
+		final HashSet<Integer> bpartnerLocationRepoIds = new HashSet<>();
+		final HashSet<Integer> bpartnerContactRepoIds = new HashSet<>();
+		final HashSet<DocTypeId> docTypeIds = new HashSet<>();
+		final HashSet<WarehouseId> warehouseIds = new HashSet<>();
+		final HashSet<CalendarId> calendarIds = new HashSet<>();
+		final HashSet<YearId> yearIds = new HashSet<>();
+
+		for (final ModelToIndex request : requests)
+		{
+			final TableRecordReference sourceModelRef = request.getSourceModelRef();
+			if (sourceModelRef == null)
+			{
+				continue;
+			}
+
+			final String sourceTableName = sourceModelRef.getTableName();
+			if (I_C_Invoice.Table_Name.equals(sourceTableName))
+			{
+				invoiceIds.add(InvoiceId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_BPartner.Table_Name.equals(sourceTableName))
+			{
+				bPartnerIds.add(BPartnerId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_BPartner_Location.Table_Name.equals(sourceTableName))
+			{
+				bpartnerLocationRepoIds.add(sourceModelRef.getRecord_ID());
+			}
+			else if (I_AD_User.Table_Name.equals(sourceTableName))
+			{
+				bpartnerContactRepoIds.add(sourceModelRef.getRecord_ID());
+			}
+			else if (I_C_DocType.Table_Name.equals(sourceTableName))
+			{
+				docTypeIds.add(DocTypeId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_M_Warehouse.Table_Name.equals(sourceTableName))
+			{
+				warehouseIds.add(WarehouseId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_Calendar.Table_Name.equals(sourceTableName))
+			{
+				calendarIds.add(CalendarId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+			else if (I_C_Year.Table_Name.equals(sourceTableName))
+			{
+				yearIds.add(YearId.ofRepoId(sourceModelRef.getRecord_ID()));
+			}
+		}
+
+		final Set<InvoiceQuery> invoiceQueries = new HashSet<>();
+
+		if (!invoiceIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().invoiceIds(invoiceIds).build());
+		}
+		if (!bPartnerIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().bpartnerIds(bPartnerIds).build());
+		}
+		if (!bpartnerLocationRepoIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().bpartnerLocationRepoIds(bpartnerLocationRepoIds).build());
+		}
+		if (!bpartnerContactRepoIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().bpartnerContactRepoIds(bpartnerContactRepoIds).build());
+		}
+		if (!warehouseIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().warehouseIds(warehouseIds).build());
+		}
+		if (!docTypeIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().docTypeIds(docTypeIds).build());
+		}
+		if (!calendarIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().calendarIds(calendarIds).build());
+		}
+		if (!yearIds.isEmpty())
+		{
+			invoiceQueries.add(InvoiceQuery.builder().yearIds(yearIds).build());
+		}
+
+		return invoiceDAO.retrieveInvoiceIds(InvoiceMultiQuery.builder().queries(invoiceQueries).build());
```

Comment:
what shall happen if invoiceQueries  is empty?
(or we don't have that case?)

Reply:
I could only see this case to happen if new entries are added ES_FTS_Config_SourceModel without being added in this code. Thrown exception to avoid this.